analyzer/emerald/emerald.go

```diff
@@ -270,23 +270,29 @@ func (m *Main) prepareBlockData(ctx context.Context, round uint64, batch *storag
 		return err
 	}
 
-	for _, f := range []func(*storage.QueryBatch, *storage.RuntimeBlockData) error{
+	for _, f := range []func(context.Context, *storage.QueryBatch, *storage.RuntimeBlockData) error{
 		m.queueBlockAndTransactionInserts,
 	} {
-		if err := f(batch, data); err != nil {
+		if err := f(ctx, batch, data); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
-func (m *Main) queueBlockAndTransactionInserts(batch *storage.QueryBatch, data *storage.RuntimeBlockData) error {
+func (m *Main) queueBlockAndTransactionInserts(ctx context.Context, batch *storage.QueryBatch, data *storage.RuntimeBlockData) error {
 	blockData, err := extractRound(data.BlockHeader, data.TransactionsWithResults, m.logger)
 	if err != nil {
 		return fmt.Errorf("extract round %d: %w", data.Round, err)
 	}
 
+	// todo: move out of queue-* section
+	blockTokenData, err := downloadRoundTokens(ctx, m.logger, m.cfg.Source.(*oasis.RuntimeClient), data.Round, blockData.DirtyTokens)
```

Comment:
we had some ideas about doing this in a separate analyzer. the emerald scanner would record any tokens it encounters, then the other analyzer would read that list and run these online calls, then store the results similar to off-chain data

Reply:
oh also we gotta unhack that downcast :skull:

Reply:
k

Reply:
hm moved to modules/evm.go. let's see how the team feels. it's a lot of exported symbols compared to most modules currently